### ISSUE TITLE
Enable Link-Time Optimization (LTO)

### ISSUE THREAD

**Issue by zamazan4ik** [pier-cli/pier#97]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit (not critical at all for this tool but anyway).

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I have made quick local tests (Fedora 40, the latest `pier` sources version): enabling `lto = true` dropped the `pier` binary size from 1.8 Mib to 1.4 Mib.

Thank you.

P.S. It's more like an improvement idea rather than a bug. I created the issue just because the Discussions are disabled for the repo for now.

**Comment by BenSchZA** [comment]
Hey @zamazan4ik! Thanks for the idea. I see no issue with enabling this for release builds. Did you see any performance improvement when you enabled this (out of curiosity)? Please feel free to submit a PR with the necessary changes, so that you'll be acknowledged as a contributor :)

**Comment by zamazan4ik** [comment]
> Did you see any performance improvement when you enabled this (out of curiosity)?

Nope, I didn't since I even didn't try to measure it properly - designing of such a bench suite will take some time ;) For Pier, the biggest improvement will be the binary size (also an important thing!)

> Please feel free to submit a PR with the necessary changes, so that you'll be acknowledged as a contributor :)

Sure! Here we go: https://github.com/pier-cli/pier/pull/98
